src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,173 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.Objects;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    public class MyNode<T> {
+        private T value;
+        private MyNode<T> prev;
+        private MyNode<T> next;
+
+        MyNode(T x) {
+            value = x;
+        }
+    }
+
+    private MyNode head;
+    private MyNode tail;
+    private int size;
+
+    public MyLinkedList() {
+        head = null;
+        tail = null;
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
+        MyNode myNode = new MyNode(value);
+        if (size == 0) {
+            head = myNode;
+            tail = myNode;
+        } else {
+            tail.next = myNode;
+            myNode.prev = tail;
+            tail = myNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Out of bounds");
+        }
+
+        MyNode myNode = new MyNode(value);
+
+        if (index == size) {
+
+            add(value);
+        } else if (index == 0) {
+            myNode.next = head;
+            head.prev = myNode;
+            head = myNode;
+            size++;
+        } else {
+
+            MyNode current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+            myNode.next = current;
+            myNode.prev = current.prev;
+            current.prev.next = myNode;
+            current.prev = myNode;
+            size++;
+        }
+
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
+
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Out of bounds");
+        }
+
+        MyNode current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return (T) current.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+
+        MyNode current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
```

Comment:
Consider creating a `findNodeByIndex` method to avoid code duplication in node traversal. This pattern is repeated in `get`, `set`, and `remove` methods.